### PR TITLE
🌵 Complete optimum conversion/merge configs

### PR DESCRIPTION
**Pull request by trajepl**
## Describe your changes

Complete optimum conversion/merge configs.
Take conversion as example, the `fp16=True` means to load the torch model with float16 from huggingface model hub. Then the fp16 torch model is converted to onnx model. 
That is different the fp16 in transformer optimization.

## Checklist before requesting a review
- [x] Add unit tests for this change.
- [x] Make sure all tests can pass.
- [x] Update documents if necessary.
- [x] Lint and apply fixes to your code by running `lintrunner -a`
- [x] Is this a user-facing change? If yes, give a description of this change to be included in the release notes.

### CODE REVIEW

File: olive/passes/onnx/optimum_conversion.py [hunk 1]
```diff
@@ -21,10 +21,29 @@ class OptimumConversion(Pass):
...
             "target_opset": PassConfigParam(
                 type_=int, default_value=14, description="The version of the default (ai.onnx) opset to target."
-            )
+            ),
+            "fp16": PassConfigParam(
+                type_=bool,
+                default_value=False,
+                description="Whether to use fp16 precision to load torch model and then convert it to onnx.",
+            ),
+            "device": PassConfigParam(
```
Comment (jambayk) [inline olive/passes/onnx/optimum_conversion.py:34]: Should we set the default device to be `None` (or cpu) like with the conversion pass and make the user specify their own device if needed? Just like in conversion, we might want have gpu target but do conversion on CPU due to memory limits, etc.

Comment (jambayk) [reply olive/passes/onnx/optimum_conversion.py:34]: Conversion on cpu still seems to be recommended. GPU is only required for fp16

File: olive/passes/onnx/optimum_conversion.py [hunk 2]
```diff
@@ -21,10 +21,29 @@ class OptimumConversion(Pass):
 
     @staticmethod
     def _default_config(accelerator_spec: AcceleratorSpec) -> Dict[str, PassConfigParam]:
+        device = "cuda" if accelerator_spec.accelerator_type == "gpu" else "cpu"
         config = {
             "target_opset": PassConfigParam(
                 type_=int, default_value=14, description="The version of the default (ai.onnx) opset to target."
-            )
+            ),
+            "fp16": PassConfigParam(
```
Comment (jambayk) [inline olive/passes/onnx/optimum_conversion.py:29]: We could add a search point validation so that `fp16` can only be true if `"cuda" in device`.

File: olive/passes/onnx/optimum_conversion.py [hunk 3]
```diff
@@ -41,22 +60,25 @@ def _run_for_config(
...
-            export_optimum_model(
-                model.model_path or hf_config.model_name,
-                output_model_path,
-                opset=config["target_opset"],
-                legacy=True,
-                no_post_process=True,
-            )
+        if version.parse(optimum_version.__version__) >= version.parse("1.14.0"):
+            # Optimum 1.14.0 needs to be run in legacy mode to support older versions of transformers
+            config["extra_args"]["legacy"] = True
```
Comment (jambayk) [inline olive/passes/onnx/optimum_conversion.py:75]: do you have plans to deprecate legacy support or make it default to `False`. Since optimum is planning to deprecate legacy mode, I think we should have legacy as opt-in instead of the default.

File: olive/passes/onnx/optimum_conversion.py (hunk 2 above)
Comment (trajepl) [reply olive/passes/onnx/optimum_conversion.py:31]: Done.

File: olive/passes/onnx/optimum_conversion.py (hunk 3 above)
Comment (trajepl) [reply olive/passes/onnx/optimum_conversion.py:90]: Yes, we should deprecate the legacy way by default and provide the options for users to enable for some corner cases in the future. 
I remembered that when I test the way with legacy=False for optimum 1.14.0, the open llama case failed for some reasons. 

Added the TODO item, will try to investigate the failure and remove legacy once the validations pass.

File: olive/passes/onnx/optimum_conversion.py [hunk 4]
```diff
@@ -24,11 +27,41 @@ def _default_config(accelerator_spec: AcceleratorSpec) -> Dict[str, PassConfigPa
...
         config.update(get_external_data_config())
         return config
 
+    def validate_search_point(
+        self, search_point: Dict[str, Any], accelerator_spec: AcceleratorSpec, with_fixed_value: bool = False
+    ) -> bool:
+        if with_fixed_value:
+            search_point = self.config_at_search_point(search_point or {})
+
+        if search_point.get("device") == "cuda" and not search_point.get("fp16"):
```
Comment (jambayk) [inline olive/passes/onnx/optimum_conversion.py:59]: I was thinking of the other direction, `if search_point.get("fp16") and "cuda" not in search_point.get("device")`. (or device=="cpu")
Similar to https://github.com/microsoft/Olive/blob/main/olive/passes/onnx/conversion.py#L92.

Cuda is required for fp16 but not the other way around. We can still do fp32 on cuda.

File: olive/passes/onnx/optimum_conversion.py (hunk 3 above)
Comment (jambayk) [reply olive/passes/onnx/optimum_conversion.py:90]: Thanks! this sounds good to me.

File: olive/passes/onnx/optimum_conversion.py (hunk 4 above)
Comment (trajepl) [reply olive/passes/onnx/optimum_conversion.py]: Oh, yes. Updated now! Thanks!

Comment (jambayk) [reply olive/passes/onnx/optimum_conversion.py]: `search_point.get("device") == "cpu" and search_point.get("fp16")` is cleaner. optimum does the same https://github.com/huggingface/optimum/blob/main/optimum/exporters/onnx/__main__.py#L290

Comment (jambayk) [reply olive/passes/onnx/optimum_conversion.py]: Oh just saw your update. I think your check is fine too. Unless we want to support something like `"cuda:0"`

**Review by jambayk — APPROVED** [review]

File: olive/passes/onnx/optimum_conversion.py (hunk 4 above)
Comment (trajepl) [reply olive/passes/onnx/optimum_conversion.py]: oh, cool. BTW, I previously try to add device id into Olive, but close that PR. There is a simple way to achieve that:
`CUDA_VISIBLE_DEVICES=1,2 python -m olive.workflows.run --config xxx`

File: test/unit_test/passes/onnx/test_optimum_conversion.py [hunk 5]
```diff
@@ -12,14 +12,45 @@
...
+)
+def test_optimum_configs(config, is_valid, tmp_path):
+    input_model = get_optimum_model_by_hf_config()
+    # setup
+    p = create_pass_from_dict(OptimumConversion, config, disable_search=True)
+    output_folder = tmp_path
+
+    if not is_valid:
+        assert p.validate_search_point(config, None) is False
+        with pytest.raises(
```
Comment (jambayk) [inline test/unit_test/passes/onnx/test_optimum_conversion.py:48]: nit: is this part needed? 

I think matching a specific error message makes the test breakable if optimum changes the error message. Or is this the purpose? So that we can catch any incompatibility with new optimum changes?

Comment (trajepl) [reply test/unit_test/passes/onnx/test_optimum_conversion.py:48]: I also try to remove the match part in my first version of implementation. But as you said, once they change the code and have different supports on fp16, then we can catch that.

**Review by jambayk — APPROVED** [review]
